src/obfs/auth_chain.c

```diff
@@ -207,21 +219,405 @@ unsigned int auth_chain_a_get_rand_len(int datalength, shift128plus_ctx *random,
+              special_data->data_size_list0_length,
+              sizeof(int),
+              data_size_list_compare
+        );
+    }
+
+    free(random);
+}
+
+void auth_chain_f_init_data_size(obfs *self, server_info_t *server, const uint8_t *key_change_datetime_key_bytes) {
+
+    auth_chain_c_data *special_data = (auth_chain_c_data *)
+            ((auth_chain_local_data *) self->l_data)->auth_chain_special_data;
+
+    shift128plus_ctx *random = (shift128plus_ctx *) malloc(sizeof(shift128plus_ctx));
+
+    uint8_t *newKey = (uint8_t *) malloc(sizeof(uint8_t) * server->key_len);
+    memcpy(newKey, server->key, server->key_len);
+    for (int i = 0; i != 8; ++i) {
+        newKey[i] ^= key_change_datetime_key_bytes[i];
+    }
+
+    shift128plus_init_from_bin(random, newKey, 16);
+    free(newKey);
+    newKey = NULL;
+
+    special_data->data_size_list0_length = shift128plus_next(random) % (8 + 16) + (4 + 8);
+    special_data->data_size_list0 = (int *) malloc(AUTH_CHAIN_D_MAX_DATA_SIZE_LIST_LIMIT_SIZE * sizeof(int));
+    for (int i = 0; i < special_data->data_size_list0_length; i++) {
+        special_data->data_size_list0[i] = shift128plus_next(random) % 2340 % 2040 % 1440;
+    }
+    // stdlib qsort
+    qsort(special_data->data_size_list0,
+          special_data->data_size_list0_length,
+          sizeof(int),
+          data_size_list_compare
+    );
+
+    int old_len = special_data->data_size_list0_length;
+    auth_chain_d_check_and_patch_data_size(self, random);
+    if (old_len != special_data->data_size_list0_length) {
+        // if check_and_patch_data_size are work, re-sort again.
+        // stdlib qsort
+        qsort(special_data->data_size_list0,
+              special_data->data_size_list0_length,
+              sizeof(int),
+              data_size_list_compare
+        );
+    }
+
+    free(random);
+}
+
+void *auth_chain_a_init_data()
+{
+    auth_chain_global_data *global = (auth_chain_global_data *) malloc(sizeof(auth_chain_global_data));
+    rand_bytes(global->local_client_id, 4);
+    rand_bytes((uint8_t *) &global->connection_id, 4);
+    global->connection_id &= 0xFFFFFF;
+    return global;
+}
+
+void *auth_chain_b_init_data() {
+    return auth_chain_a_init_data();
+}
+
+void *auth_chain_c_init_data() {
+    return auth_chain_a_init_data();
+}
+
+void *auth_chain_d_init_data() {
+    return auth_chain_c_init_data();
+}
+
+void *auth_chain_e_init_data() {
+    return auth_chain_d_init_data();
+}
+
+void *auth_chain_f_init_data() {
+    return auth_chain_e_init_data();
+}
+
 obfs *auth_chain_a_new_obfs()
 {
     obfs *self = new_obfs();
     self->l_data = malloc(sizeof(auth_chain_local_data));
     auth_chain_local_data_init((auth_chain_local_data *) self->l_data);
     ((auth_chain_local_data *) self->l_data)->salt = "auth_chain_a";
-    ((auth_chain_local_data *) self->l_data)->rnd = auth_chain_a_get_rand_len;
+    ((auth_chain_local_data *) self->l_data)->get_tcp_rand_len = auth_chain_a_get_rand_len;
```

Comment:
> ../src/obfs/auth_chain.c: In function ‘auth_chain_a_new_obfs’:
../src/obfs/auth_chain.c:544:64: warning: assignment to ‘unsigned int (*)(struct auth_chain_local_data *, server_info_t *, int,  shift128plus_ctx *, uint8_t *)’ {aka ‘unsigned int (*)(struct auth_chain_local_data *, struct <anonymous> *, int,  struct shift128plus_ctx *, unsigned char *)’} from incompatible pointer type ‘unsigned int (*)(int,  shift128plus_ctx *, uint8_t *, int *, int,  int *, int,  int)’ {aka ‘unsigned int (*)(int,  struct shift128plus_ctx *, unsigned char *, int *, int,  int *, int,  int)’} [-Wincompatible-pointer-types]
  544 |     ((auth_chain_local_data *) self->l_data)->get_tcp_rand_len = auth_chain_a_get_rand_len;
      |